### ISSUE TITLE
Improve relations logic

### ISSUE THREAD

**Issue by crobinson42** [js-data/js-data#558]
## Description

Js-data provides the ability to define relations between Mappers. In v3 there are a few issues that have been reported that [need attention ](https://github.com/js-data/js-data/labels/mapper%20relations). From experience in helping js-data users in the Slack channels, I have noticed that the relations definitions and setup doesn't seem so intuitive to a lot of folks and there are some gotchas that we could work towards improving.

Relations also need better unit tests to ensure their stability.

Also, the documentation on the site is not thorough regarding how to setup and the use cases for relations.

**Comment by zeesulehria** [comment]
@crobinson42 There is also need for inheritance at relationship level. Currently we need to duplicate all relation definition for subtypes which is a major issue.

**Comment by crobinson42** [comment]
@zeesulehria this would need contribution from the community -  if you're willing to help with the efforts let me know :-)

**Comment by zeesulehria** [comment]
@crobinson42 I'd be willing to contribute to take this framework forward.

**Comment by crobinson42** [comment]
@zeesulehria great 👍 are you in the slack team? http://slack.js-data.io/

**Comment by zeesulehria** [comment]
@crobinson42 yes I am in js-data slack, username: `zeesulehria` Zeeshan Aslam
